### PR TITLE
EOS-20429: SW Upgrade pre validation - upgrade of CORTX components is compatible with the current version

### PR DESCRIPTION
**Pull request by dmitrydb**
I added SW upgrade packages compatibility validation.

### CODE REVIEW

File: api/python/provisioner/commands/validator/validator.py [hunk 1]
```diff
@@ -784,3 +787,128 @@ def __attrs_post_init__(self):
...
+
+        error_msg = list()
+        for pkg in iso_info.packages:
+            if (SWUpgradeInfoFields.VERSION_COMPATIBILITY.value
+                    in iso_info.packages[pkg]):
+                compatibility = iso_info.packages[pkg][
+                    SWUpgradeInfoFields.VERSION_COMPATIBILITY.value]
+
+                installed_ver = packages.get(pkg, None)
+                if installed_ver is None:
```
Comment (andkononykhin2) [inline api/python/provisioner/commands/validator/validator.py:893]: if a package is not installed it should be ok to install any new version, so it's not an error

File: api/python/provisioner/commands/validator/validator.py [hunk 2]
```diff
@@ -784,3 +787,128 @@ def __attrs_post_init__(self):
...
+        Raises
+        ------
+        ValidationError
+            If validation is failed.
+        """
+        packages = list(iso_info.packages.keys())
+
+        # NOTE: the first line of `yum -q list installed` command is
+        #  'Installed Packages' skip it via `tail -n +2`
+        cmd = (f"yum -q list installed {packages} | tail -n +2 | "
```
Comment (andkononykhin2) [inline api/python/provisioner/commands/validator/validator.py:860]: edge case: packages list is empty

File: api/python/provisioner/commands/validator/validator.py [hunk 3]
```diff
@@ -784,3 +787,128 @@ def __attrs_post_init__(self):
...
+        Returns
+        -------
+        None
+
+        Raises
+        ------
+        ValidationError
+            If validation is failed.
+        """
+        packages = list(iso_info.packages.keys())
```
Comment (andkononykhin2) [inline api/python/provisioner/commands/validator/validator.py:856]: shouldn't we check Cortx release constraint as well?

File: api/python/provisioner/commands/validator/validator.py [hunk 4]
```diff
@@ -784,3 +787,128 @@ def __attrs_post_init__(self):
...
+@attr.s(auto_attribs=True)
+class CompatibilityValidator:
+
+    """
+    Validator that checks if all SW upgrade packages are compatible with
+    installed ones.
+    """
+
+    @staticmethod
+    def _compare_versions(a: str, b: str, operator: str) -> bool:
```
Comment (andkononykhin2) [inline api/python/provisioner/commands/validator/validator.py:801]: that custom implemetation looks good, but I would suggest to simplify our life and use [packaging](https://packaging.pypa.io/en/latest/specifiers.html) here

```bash
>>> from packaging.specifiers import Version, SpecifierSet
>>> old_v = Version('1.0.0-321')
>>> new_v = Version('2.0.0-321')
>>> constraint = SpecifierSet('> 2.0.0-0')
>>> old_v in constraint
False
>>> new_v in constraint
True
>>> old_v2 = Version('2.0.0-0')
>>> old_v2 in constraint
False
```

**Review by andkononykhin2 — COMMENTED** [review]
@dmitrydb Looks good. I have one major comment regarding not existent pkg error. We shouldn't fail for that case. Also please consider to use packaging library for versions comparison. Thanks

File: api/python/provisioner/commands/validator/validator.py (hunk 4 above)
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py]: Yes, I rewrote this part using the proposed functionality, thank you! I didn't expect that `packaging` supports such API.

File: api/python/provisioner/commands/validator/validator.py (hunk 1 above)
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py:868]: Done.

File: api/python/provisioner/commands/validator/validator.py (hunk 2 above)
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py]: It is not a problem, but it is better to stop validation in this case to save the CPU time. Done

File: api/python/provisioner/commands/validator/validator.py (hunk 3 above)
Comment (dmitrydb) [reply api/python/provisioner/commands/validator/validator.py:834]: I added that validation for CORTX version

**Review by andkononykhin2 — APPROVED** [review]

**Pull request merged**